experimental/ast/decl_def.go

```diff
@@ -0,0 +1,541 @@
+// there is one.
+//
+// See [DeclDef.Type] for details on where this keyword comes from.
+func (d DeclDef) Keyword() Token {
+	path, ok := d.Type().(TypePath)
+	if !ok {
+		return Token{}
+	}
+	ident := path.Path.AsIdent()
+	switch ident.Text() {
+	case "message", "enum", "service", "extend", "oneof", "group", "rpc", "option":
+		return ident
+	default:
+		return Token{}
+	}
+}
+
+// Name returns this definition's declared name.
+func (d DeclDef) Name() Path {
+	return d.raw.name.With(d)
+}
+
+// Signature returns this definition's type signature, if it has one.
+//
+// Note that this is distinct from the type returned by [DeclDef.Type], which
+// is the "prefix" type for the definition (such as for a field). This is a
+// signature for e.g. a method.
+//
+// Not all defs have a signature, so this function may return a nil Signature.
+// If you want to add one, use [DeclDef.WithSignature].
+func (d DeclDef) Signature() Signature {
+	if d.raw.signature == nil {
+		return Signature{}
+	}
+
+	return Signature{
+		d.withContext,
+		d.raw.signature,
+	}
+}
+
+// WithSignature is like Signature, but it adds an empty signature if it would
+// return nil.
+func (d DeclDef) WithSignature() Signature {
+	if d.Signature().Nil() {
+		d.raw.signature = new(rawSignature)
+	}
+	return d.Signature()
+}
+
+// Equals returns this definitions = token, before the value.
+// May be nil.
+func (d DeclDef) Equals() Token {
+	return d.raw.equals.With(d)
+}
+
+// Value returns this definition's value. For a field, this will be the
+// tag number, while for an option, this will be the complex expression
+// representing its value.
+func (d DeclDef) Value() Expr {
+	return d.raw.value.With(d)
+}
+
+// SetValue sets the value of this definition.
+//
+// See [DeclDef.Value].
+func (d DeclDef) SetValue(expr Expr) {
+	d.raw.value = toRawExpr(expr)
+}
+
+// Options returns the compact options list for this definition.
+func (d DeclDef) Options() CompactOptions {
+	return newOptions(d.raw.options, d)
+}
+
+// SetOptions sets the compact options list for this definition.
+//
+// Setting it to a nil Options clears it.
+func (d DeclDef) SetOptions(opts CompactOptions) {
+	d.raw.options = opts.ptr
+}
+
+// Body returns this definition's body, if it has one.
+func (d DeclDef) Body() DeclScope {
+	return wrapDecl[DeclScope](arena.Untyped(d.raw.body), d)
+}
+
+// SetBody sets the body for this definition.
+func (d DeclDef) SetBody(b DeclScope) {
+	d.raw.body = arena.Pointer[rawDeclScope](b.ptr)
+}
+
+// Semicolon returns the ending semicolon token for this definition.
+// May be nil.
+func (d DeclDef) Semicolon() Token {
+	return d.raw.semi.With(d)
+}
+
+// Classify looks at all the fields in this definition and decides what kind of
+// definition it's supposed to represent.
+//
+// For nonsensical definitions, this returns nil, although it is not guaranteed
+// to return nil for *all* nonsensical definitions.
+func (d DeclDef) Classify() Def {
```

Comment:
Need to think about this. You will be pleased/horrified to learn that the legalizer does not actually call Classifier; it needs to do classification itself, even more carefully than Classify does, because it needs to diagnose things carefully.

I am also not sure if I think this is the right interface/contract for Classify... I expect it to evolve as we work on the parser stack together.